internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	prefixedKey := prependPrefix(pDB.prefix, key)
+	if err := pDB.db.DeleteSync(prefixedKey); err != nil {
+		return fmt.Errorf("prefixed DB namespace synced delete: %w", err)
+	}
+
+	return nil
+}
+
+// Iterator returns an iterator over a domain of keys, in ascending order.
+// The caller must call [Close] when done. End is exclusive, and start must be
+// less than end. A nil start iterates from the first key, and a nil end
+// iterates to the last key (inclusive). Empty keys are not valid.
+// No writes may happen within a domain while an iterator exists over it.
+//
+// Do not modify the contents of the arguments while the returned iterator is in use.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Iterator(start, end []byte) (Iterator, error) {
+	itStart, itEnd, err := prefixedIteratorBounds(pDB.prefix, start, end)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	it, err := pDB.db.Iterator(itStart, itEnd)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace iterator: %w", err)
+	}
+
+	return newPrefixDBIterator(pDB.prefix, start, end, it)
+}
+
+// ReverseIterator returns an iterator over a domain of keys, in descending
+// order. The caller must call Close when done. End is exclusive, and start must
+// be less than end. A nil end iterates from the last key (inclusive), and a nil
+// start iterates to the first key (inclusive). Empty keys are not valid.
+// No writes may happen within a domain of keys while an iterator exists over it.
+//
+// Do not modify the contents of the arguments while the returned iterator is in use.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) ReverseIterator(start, end []byte) (Iterator, error) {
+	itStart, itEnd, err := prefixedIteratorBounds(pDB.prefix, start, end)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	it, err := pDB.db.ReverseIterator(itStart, itEnd)
+	if err != nil {
+		return nil, fmt.Errorf("prefixed DB namespace reverse iterator: %w", err)
+	}
+
+	return newPrefixDBIterator(pDB.prefix, start, end, it)
+}
+
+// NewBatch creates a batch for atomic database updates.
+// The caller is responsible for calling Batch.Close() once done.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) NewBatch() Batch {
+	pDB.mu.Lock()
+	defer pDB.mu.Unlock()
+
+	return newPrefixDBBatch(pDB.prefix, pDB.db.NewBatch())
+}
+
+// Compact compacts the specified range of keys in the database.
+//
+// It implements the [DB] interface for type PrefixDB.
+func (pDB *PrefixDB) Compact(start, end []byte) error {
+	if err := pDB.db.Compact(start, end); err != nil {
+		return fmt.Errorf("prefixed DB namespace compaction: %w", err)
```

Comment:
`Compact` does not prepend the prefix to `start` and `end` before calling `Compact` on the underlying database. Is this intended or an oversight?

Given the description of `PrefixDB`, I would expect that **all** read and write operations prepend the prefix to the keys. While this means it would only compact keys with the same prefix, which might not be particularly useful, I think not prepending the prefix goes against the `PrefixDB` type documentation.

Reply:
Good catch! This was probably an oversight. @jmalicevic would know more

Reply:
Oh yes, definitely an oversight. Although the way we use it now it did not cause problems because we pass `nil,nil` as the argument to indicate we want everything compacted. Although tbh I am not sure how this is handled then, if the keys are `nil,nil` we should most likely ignore the prefix. This will just try to compact all the files, it does not delete or alter them in any way. Albeit it might slow things down.